From kosaki.motohiro@jp.fujitsu.com  Thu Apr  9 14:36:15 2009
From: KOSAKI Motohiro <kosaki.motohiro@jp.fujitsu.com>
Date: Thu,  9 Apr 2009 13:57:59 +0900 (JST)
Subject: proc: mounts_poll() make consistent to mdstat_poll
To: Neil Brown <neilb@suse.de>, Greg Kroah-Hartman <gregkh@suse.de>, Al Viro <viro@ZenIV.linux.org.uk>, Ram Pai <linuxram@us.ibm.com>, Miklos Szeredi <mszeredi@suse.cz>
Cc: kosaki.motohiro@jp.fujitsu.com
Message-ID: <20090409135633.B3E9.A69D9226@jp.fujitsu.com>


In recently sysfs_poll discussion, Neil Brown pointed out /proc/mounts
also should be fixed.

SUSv3 says "Regular files shall always poll TRUE for reading and
writing".  see
http://www.opengroup.org/onlinepubs/009695399/functions/poll.html

Then, mounts_poll()'s default should be "POLLIN | POLLRDNORM".  it mean
always readable.

In addition, event trigger should use "POLLERR | POLLPRI" instead
POLLERR.  it makes consistent to mdstat_poll() and sysfs_poll(). and,
select(2) can handle POLLPRI easily.


Reported-by: Neil Brown <neilb@suse.de>
Signed-off-by: KOSAKI Motohiro <kosaki.motohiro@jp.fujitsu.com>
Cc: Ram Pai <linuxram@us.ibm.com>
Cc: Miklos Szeredi <mszeredi@suse.cz>
Cc: Al Viro <viro@zeniv.linux.org.uk>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 fs/proc/base.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/fs/proc/base.c
+++ b/fs/proc/base.c
@@ -648,14 +648,14 @@ static unsigned mounts_poll(struct file 
 {
 	struct proc_mounts *p = file->private_data;
 	struct mnt_namespace *ns = p->ns;
-	unsigned res = 0;
+	unsigned res = POLLIN | POLLRDNORM;
 
 	poll_wait(file, &ns->poll, wait);
 
 	spin_lock(&vfsmount_lock);
 	if (p->event != ns->event) {
 		p->event = ns->event;
-		res = POLLERR;
+		res |= POLLERR | POLLPRI;
 	}
 	spin_unlock(&vfsmount_lock);
 
